adam/learner/cross_situational.py

```diff
@@ -0,0 +1,567 @@
+import logging
+from pathlib import Path
+from random import Random
+from typing import Dict, Generic, Iterable, List, Mapping, Optional, Set, Tuple
+
+from attr.validators import in_, instance_of, optional
+from immutablecollections import immutabledict
+from more_itertools import first
+from vistautils.parameters import Parameters
+from vistautils.range import Range
+
+from adam.language import (
+    LinguisticDescription,
+    LinguisticDescriptionT,
+    TokenSequenceLinguisticDescription,
+)
+from adam.learner import (
+    LanguageLearner,
+    LearningExample,
+    get_largest_matching_pattern,
+    graph_without_learner,
+)
+from adam.ontology.ontology import Ontology
+from adam.ontology.phase1_ontology import GAILA_PHASE_1_ONTOLOGY
+from adam.ontology.phase1_spatial_relations import Region
+from adam.perception import ObjectPerception, PerceptionT, PerceptualRepresentation
+from adam.perception.developmental_primitive_perception import (
+    DevelopmentalPrimitivePerceptionFrame,
+    RgbColorPerception,
+)
+from adam.perception.perception_graph import (
+    DebugCallableType,
+    GraphLogger,
+    PerceptionGraph,
+    PerceptionGraphPattern,
+)
+from adam.utils import networkx_utils
+from attr import Factory, attrib, attrs
+
+
+@attrs
+class CrossSituationalLanguageLearner(
+    Generic[PerceptionT, LinguisticDescriptionT],
+    LanguageLearner[PerceptionT, LinguisticDescription],
+):
+    """
+    An implementation of `LanguageLearner` for cross-situational learning based approach for single
+    object detection.
+    """
+
+    _words_to_hypotheses_and_scores: Dict[
+        str, Dict[PerceptionGraphPattern, float]
+    ] = attrib(init=False, default=Factory(dict))
+    _words_to_hypotheses_and_probability: Dict[
```

Comment:
It would be useful to have a comment here describing the difference between score and probability